src/telemetry/telemetry.ts

```diff
@@ -7,56 +7,92 @@ import { MACHINE_METADATA } from "./constants.js";
+        await fs.access(filePath, fs.constants.F_OK);
+        return true; // File exists
+    } catch (e: unknown) {
+        if (
+            e instanceof Error &&
+            (
+                e as Error & {
+                    code: string;
+                }
+            ).code === "ENOENT"
+        ) {
+            return false; // File does not exist
+        }
+        throw e; // Re-throw unexpected errors
+    }
+}
 
-type EventResult = {
-    success: boolean;
-    error?: Error;
-};
+async function isContainerized(): Promise<boolean> {
+    if (process.env.container) {
+        return true;
+    }
+
+    const exists = await Promise.all(["/.dockerenv", "/run/.containerenv", "/var/run/.containerenv"].map(fileExists));
 
-export const DEVICE_ID_TIMEOUT = 3000;
+    return exists.includes(true);
+}
 
 export class Telemetry {
-    private isBufferingEvents: boolean = true;
-    /** Resolves when the device ID is retrieved or timeout occurs */
-    public deviceIdPromise: Promise<string> | undefined;
+    private deviceId: string | undefined;
+    private containerEnv: boolean | undefined;
     private deviceIdAbortController = new AbortController();
     private eventCache: EventCache;
     private getRawMachineId: () => Promise<string>;
+    private getContainerEnv: () => Promise<boolean>;
 
     private constructor(
         private readonly session: Session,
         private readonly userConfig: UserConfig,
-        private readonly commonProperties: CommonProperties,
-        { eventCache, getRawMachineId }: { eventCache: EventCache; getRawMachineId: () => Promise<string> }
+        {
+            eventCache,
+            getRawMachineId,
+            getContainerEnv,
+        }: {
+            eventCache: EventCache;
+            getRawMachineId: () => Promise<string>;
+            getContainerEnv: () => Promise<boolean>;
+        }
     ) {
         this.eventCache = eventCache;
         this.getRawMachineId = getRawMachineId;
+        this.getContainerEnv = getContainerEnv;
     }
 
-    static create(
+    static async create(
         session: Session,
         userConfig: UserConfig,
         {
-            commonProperties = { ...MACHINE_METADATA },
             eventCache = EventCache.getInstance(),
             getRawMachineId = () => nodeMachineId.machineId(true),
+            getContainerEnv = isContainerized,
         }: {
             eventCache?: EventCache;
             getRawMachineId?: () => Promise<string>;
-            commonProperties?: CommonProperties;
+            getContainerEnv?: () => Promise<boolean>;
         } = {}
-    ): Telemetry {
-        const instance = new Telemetry(session, userConfig, commonProperties, { eventCache, getRawMachineId });
+    ): Promise<Telemetry> {
+        const instance = new Telemetry(session, userConfig, {
+            eventCache,
+            getRawMachineId,
+            getContainerEnv,
+        });
 
-        void instance.start();
+        await instance.start();
```

Comment:
```suggestion
        void instance.start();
```
This was an important detail in the existing implementation: **start does was not getting awaited before**. This is why the behavior is equivalent to what was trying to be accomplished before with skipping the promise. Telemetry does not wait for the device ID resolution.

So this function does not need to be asynchronous and can be left as is.